README.md

```diff
@@ -2,7 +2,141 @@
+
+You can register interceptors to mock Jenkins commands, which may or may not return a result.
+
+```groovy
+    @Override
+    @Before
+    void setUp() throws Exception {
+        super.setUp()
+        helper.registerAllowedMethod("timeout", [Map.class, Closure.class], null)
+        helper.registerAllowedMethod(method("readFile", String.class), { file ->
+            return Files.contentOf(new File(file), Charset.forName("UTF-8"))
+        })
+    }
+```
+
+The test helper already includes some mocks, but the list is far from complete. 
+You need to _register allowed methods_ if you want to override these mock and add others. 
+Note that you need to provide a method signature and a callback (closure or lambda) in order to allow a method.
+Any method call which is not recognized will throw an exception.
+
+Some tricky methods such as `load` and `parallel` are implemented directly in the helper. 
+If you want to override those, make sure that you extend the `PipelineTestHelper` class.
+
+### Analyze the mock execution
+
+The helper registers every method call to provide a stacktrace of the mock execution.
+ 
+```groovy
+
+@Test
+void should_execute_without_errors() throws Exception {
+    loadScript("Jenkinsfile")
+    assertThat(helper.callStack.findAll { call ->
+        call.methodName == "sh"
+    }.any { call ->
+        callArgsToString(call).contains("mvn verify")
+    }).isTrue()
+}
+
+```
+
+## Configuration
+
+The abstract class `BasePipelineTest` configures the helper with useful conventions: 
+
+- It looks for pipeline scripts in your project in root (`./.`) and `src/main/jenkins` paths.
+- Jenkins pipelines let you load other scripts from a parent script with `load` command. 
+However `load` takes the full path relative to the project root. 
+The test helper mock successfully the `load` command to load the scripts. But for you to not think about relative paths, 
+you need to configure the path of the project where your pipeline scripts are in your project. 
+Defaults to `production/jenkins/`
+- Pipeline script extension, defaults to `.jenkins`.
+
+Overriding these default values is easy: 
+
+```groovy
+
+class TestExampleJob extends BasePipelineTest {
+
+    @Override
+    @Before
+    void setUp() throws Exception {
+        helper.baseScriptRoot = ''
+        helper.roots += 'src/main/groovy'
+        helper.extension = 'pipeline'
+        super.setUp()
+    }
+    
+}
+
+```
+
+## Note on CPS
+
+If you already fiddled with Jenkins pipeline DSL, you experienced strange errors during execution on Jenkins.
+This is because Jenkins does not directly execute your pipeline in Groovy, 
+but transforms the pipeline code into an intermediate format to in order to run Groovy code in 
+[Continuation Passing Style](https://en.wikipedia.org/wiki/Continuation-passing_style) (CPS).
+ 
+The usual errors are partly due to the sandboxing Jenkins applies for security reasons, 
+and partly due to the serializability Jenkins imposes.
```

Comment:
Ajouter des liens:
[the sandboxing Jenkins applies](https://wiki.jenkins-ci.org/display/JENKINS/Script+Security+Plugin#ScriptSecurityPlugin-GroovySandboxing) ...
[serializability Jenkins imposes](https://github.com/jenkinsci/pipeline-plugin/blob/master/TUTORIAL.md#serializing-local-variables)